File: leetgo/cpp/LCR 121.er-wei-shu-zu-zhong-de-cha-zhao-lcof/solution.cpp
// Created by k at 2024/05/11 16:16
// leetgo: dev
// https://leetcode.cn/problems/er-wei-shu-zu-zhong-de-cha-zhao-lcof/

#include <bits/stdc++.h>
#include "LC_IO.h"
using namespace std;

// @lc code=begin

class Solution {
  public:
    bool findTargetIn2DPlants(vector<vector<int>> &plants, int target)
    {
        int i = plants.size() - 1;
        int j = 0;
        for (; i >= 0 && j < plants[0].size();) {
            if (plants[i][j] == target) {
                return true;
            }
            else if (plants[i][j] < target) {
                j++;
            }
            else {
                i--;
            }
        }
        return false;
    }
};

// @lc code=end

int main()
{
    ios_base::sync_with_stdio(false);
    stringstream out_stream;

    vector<vector<int>> plants;
    LeetCodeIO::scan(cin, plants);
    int target;
    LeetCodeIO::scan(cin, target);

    Solution *obj = new Solution();
    auto res      = obj->findTargetIn2DPlants(plants, target);
    LeetCodeIO::print(out_stream, res);
    cout << "\noutput: " << out_stream.rdbuf() << endl;

    delete obj;
    return 0;
}

File: leetgo/go/LCR 121.er-wei-shu-zu-zhong-de-cha-zhao-lcof/solution.go
// Created by k at 2024/05/11 15:48
// leetgo: dev
// https://leetcode.cn/problems/er-wei-shu-zu-zhong-de-cha-zhao-lcof/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func findTargetIn2DPlants(plants [][]int, target int) bool {
	m, n := len(plants)-1, 0
	for m >= 0 && n < len(plants[0]) {
		if plants[m][n] == target {
			return true
		} else if plants[m][n] < target {
			n++
		} else {
			m--
		}
	}
	return false
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	plants := Deserialize[[][]int](ReadLine(stdin))
	target := Deserialize[int](ReadLine(stdin))
	ans := findTargetIn2DPlants(plants, target)

	fmt.Println("\noutput:", Serialize(ans))
}
